ethcore/src/snapshot/service.rs

```diff
@@ -334,12 +338,144 @@ impl Service {
 	// replace one the client's database with our own.
 	fn replace_client_db(&self) -> Result<(), Error> {
-		let our_db = self.restoration_db();
+		let migrated_blocks = self.migrate_blocks()?;
+		trace!(target: "snapshot", "Migrated {} ancient blocks", migrated_blocks);
 
-		self.db_restore.restore_db(&*our_db.to_string_lossy())?;
+		let rest_db = self.restoration_db();
+		self.client.restore_db(&*rest_db.to_string_lossy())?;
 		Ok(())
 	}
 
+	// Migrate the blocks in the current DB into the new chain
+	fn migrate_blocks(&self) -> Result<usize, Error> {
+		// Count the number of migrated blocks
+		let mut count = 0;
+		let rest_db = self.restoration_db();
+
+		let cur_chain_info = self.client.chain_info();
+
+		let next_db = self.restoration_db_handler.open(&rest_db)?;
+		let next_chain = BlockChain::new(Default::default(), &[], next_db.clone());
+		let next_chain_info = next_chain.chain_info();
+
+		// Get the *inclusive* first and last blocks to iterate through
+		let start_block_number = match next_chain_info.ancient_block_number {
+			Some(bn) => bn + 1,
+			None => return Ok(count),
+		};
+		let mut end_block_number = match next_chain_info.first_block_number {
+			Some(bn) => bn - 1,
+			None => return Ok(count),
+		};
+
+		// Check that the intersection is non-null
+		if start_block_number >= end_block_number {
+			return Ok(count);
+		}
+
+		// Find the best block we can sync to
+		match (cur_chain_info.ancient_block_number, cur_chain_info.best_block_number) {
+			// The current chain already has some available and useful ancient blocks
+			(Some(ancient_bn), _) if ancient_bn > start_block_number => {
+				end_block_number = ancient_bn;
+			},
+			// The current chain was synced until best_bn
+			(None, best_bn) => {
+				// Take the minimum block between `best_bn` and `end_block_number`
+				if best_bn < end_block_number {
+					end_block_number = best_bn;
+				}
+			},
+			// If the ancient block number is bellow the new one, which should not happen
+			_ => return Ok(count),
+		};
+
+		// Check that the intersection is non-null
+		if start_block_number >= end_block_number {
+			return Ok(count);
+		}
+
+		// Try to include every block that will need to be downloaded from the current chain
+		// Break when no more blocks are available from it.
+		trace!(target: "snapshot", "Trying to import ancient blocks from {} to {}",
+			start_block_number, end_block_number,
+		);
+
+		// As we are going backwards, the target hash is the one of the first block's parent
+		let target_block_hash = match self.client.block_hash(BlockId::Number(start_block_number - 1)) {
+			Some(hash) => hash,
+			None => return Ok(count),
+		};
+
+		let last_block_hash = match self.client.block_hash(BlockId::Number(end_block_number)) {
+			Some(h) => h,
+			None => return Ok(count),
+		};
+
+		let mut success = false;
+		let mut batch = DBTransaction::new();
+		let mut block = self.client.block(BlockId::Hash(last_block_hash));
+
+		while let Some(block_view) = block {
```

Comment:
I'd simplify this loop a bit as well, maybe lower the number of variables and use stuff like `block_view.hash()` directly.
```rust
let mut parent_hash = last_block_hash;
while target_hash != parent_hash {
  let block = self.client.block(BlockId::Hash(parent_hash))?;
  // move to parent block
  parent_hash = block.parent_hash();

  let receipts = self.client.block_receipts(&block.hash())?;
  let parent_difficulty= self.client.block_total_difficulty(BlockId::Hash(block.parent_hash()))?;

   //insert block, etc

   if block.number() % 10_000 {
   }
}
```

^^ something like this